### ISSUE TITLE
Support for lifespan state

### ISSUE THREAD

**Issue by GuillaumeDesforges** [florimondmanca/asgi-lifespan#57]
I'm using a stateful lifetime to share a database connection on my FastAPI app, which works when started with `uvicorn`.

Since `httpx`'s `AsyncClient` does not support lifetime management, they recommend using `asgi-lifespan`.

Reference: https://www.python-httpx.org/async/#startupshutdown-of-asgi-apps

My lifetime looks like this.
```python
class State(TypedDict):
    db: Connection | Engine


@asynccontextmanager
async def lifespan(app: Starlette) -> AsyncGenerator[State, None]:
    extra = app.__dict__["extra"]
    config = extra["config"]

    engine = create_engine(url=config.database_url, echo=config.debug)

    connection: Connection | None = None

    if config.debug:
        db.Base.metadata.create_all(bind=engine)
        connection = engine.connect()

    yield {"db": connection or engine}

    if connection is not None:
        connection.close()

    return
```

then my test looks like this

```python
@pytest.mark.anyio
async def test_post_post(app: FastAPI):
    print("Start test")
    async with LifespanManager(app):
        async with AsyncClient(app=app) as client:
            response = await client.post("/post", json={"title": "test"})
            assert response.status_code == 200
            response_json = response.json()
            assert response_json["title"] == "test"
            # check it's inserted in database
            state: State = cast(State, app.state)
            with Session(state["db"]) as session:
                db_post = (
                    session.query(db.Post).filter_by(id=response_json["uid"]).first()
                )
                assert db_post is not None
                assert db_post.title == "test"
```

However I get

```
.venv\Lib\site-packages\asgi_lifespan\_manager.py:90: in __aenter__
    await self.startup()
.venv\Lib\site-packages\asgi_lifespan\_manager.py:36: in startup
    raise self._app_exception
.venv\Lib\site-packages\asgi_lifespan\_manager.py:64: in run_app
    await self.app(scope, self.receive, self.send)
.venv\Lib\site-packages\fastapi\applications.py:273: in __call__
    await super().__call__(scope, receive, send)
.venv\Lib\site-packages\starlette\applications.py:120: in __call__
    await self.middleware_stack(scope, receive, send)
.venv\Lib\site-packages\starlette\middleware\errors.py:149: in __call__
    await self.app(scope, receive, send)
.venv\Lib\site-packages\starlette\middleware\exceptions.py:55: in __call__
    await self.app(scope, receive, send)
.venv\Lib\site-packages\fastapi\middleware\asyncexitstack.py:21: in __call__
    raise e
.venv\Lib\site-packages\fastapi\middleware\asyncexitstack.py:18: in __call__
    await self.app(scope, receive, send)
.venv\Lib\site-packages\starlette\routing.py:705: in __call__
    await self.lifespan(scope, receive, send)
-----------------------------------------
self = <fastapi.routing.APIRouter object at 0x0000026E1B783ED0>
scope = {'app': <fastapi.applications.FastAPI object at 0x0000026E1B783F10>, 'fastapi_astack': <contextlib.AsyncExitStack object at 0x0000026E1CBFD890>, 'router': <fastapi.routing.APIRouter object at 0x0000026E1B783ED0>, 'type': 'lifespan'}
receive = <bound method LifespanManager.receive of <asgi_lifespan._manager.LifespanManager object at 0x0000026E1CBFD410>>
send = <bound method LifespanManager.send of <asgi_lifespan._manager.LifespanManager object at 0x0000026E1CBFD410>>

    async def lifespan(self, scope: Scope, receive: Receive, send: Send) -> None:
        """
        Handle ASGI lifespan messages, which allows us to manage application
        startup and shutdown events.
        """
        started = False
        app: typing.Any = scope.get("app")
        await receive()
        try:
            async with self.lifespan_context(app) as maybe_state:
                if maybe_state is not None:
                    if "state" not in scope:
>                       raise RuntimeError(
                            'The server does not support "state" in the lifespan scope.'
                        )
E                       RuntimeError: The server does not support "state" in the lifespan scope.

.venv\Lib\site-packages\starlette\routing.py:678: RuntimeError
```

**Comment by euri10** [comment]
was making the switch fron on_startup/on_shutdown to lifespan on some apps today and about to report the same @florimondmanca

**Comment by florimondmanca** [comment]
Hi there,

I understand something changed in the `scope` that Starlette expects, namely that there'd be some `"state"` key for it to use? Is it a standard ASGI thing, or is it something we happen to drop / not include but should?

**Comment by florimondmanca** [comment]
Using up-to-date versions of all libraries involved and Python 3.11, I can't reproduce with this sample script. What's missing to trigger the issue you're running into?

```python
import asyncio
from contextlib import asynccontextmanager
from starlette.applications import Starlette
from asgi_lifespan import LifespanManager


@asynccontextmanager
async def lifespan(app):
    print("startup")
    yield
    print("shutdown")


app = Starlette(lifespan=lifespan)


async def main():
    async with LifespanManager(app):
        print("lifespan ok")


asyncio.run(main())
```

```
startup
lifespan ok
shutdown
```

**Comment by Kludex** [comment]
There's a new section on Starlette: https://www.starlette.io/lifespan/

**Comment by florimondmanca** [comment]
My understanding of the spec is that `asgi-lifespan` (acting as an in-process / in-memory server) should ensure that ASGI scopes get a shallow copy of this dict.

https://asgi.readthedocs.io/en/latest/specs/lifespan.html#lifespan-state

> The server will ensure that a shallow copy of the namespace is passed into each subsequent request/response call into the application.

I believe this implies storing some kind of per-instance dict on `LifespanManager` instances, and passing (copying?) it when building the `{"type": "lifespan"}` scope?

I don't have the full view over the details, though, so happy to review PRs that add support for this. Greetings!

**Comment by florimondmanca** [comment]
@GuillaumeDesforges Would you like to try out #59 before we merge?


```
pip install git+https://github.com/euri10/asgi-lifespan.git@c2cf923146b03d9b04da29bf256fe90cea981527
```

**Comment by GuillaumeDesforges** [comment]
Did not work for me. See https://github.com/GuillaumeDesforges/reproduce-asgi-lifespan-57

**Comment by florimondmanca** [comment]
@GuillaumeDesforges Thanks for the reproduction repo.

On lines 10 and 11 here: https://github.com/GuillaumeDesforges/reproduce-asgi-lifespan-57/blob/master/tests/test_app.py#L10

You’d now need to do as shown in the updated README: https://github.com/florimondmanca/asgi-lifespan/pull/59/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R139

i.e. pass `app=manager.app` to AsyncClient

How do things look with that tweak?

**Comment by florimondmanca** [comment]
Released in 2.1.0 :+1: https://pypi.org/project/asgi-lifespan/2.1.0
